libvast/src/segment.cpp

```diff
@@ -0,0 +1,201 @@
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include <caf/stream_deserializer.hpp>
+
+#include <caf/detail/scope_guard.hpp>
+
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/const_table_slice_handle.hpp"
+#include "vast/ids.hpp"
+#include "vast/segment.hpp"
+#include "vast/si_literals.hpp"
+#include "vast/table_slice.hpp"
+
+#include "vast/detail/assert.hpp"
+#include "vast/detail/byte_swap.hpp"
+#include "vast/detail/coded_deserializer.hpp"
+#include "vast/detail/coded_serializer.hpp"
+#include "vast/detail/narrow.hpp"
+
+namespace vast {
+
+using namespace binary_byte_literals;
+
+namespace {
+
+template <class T>
+T to_little_endian(T x) {
+  return detail::swap<detail::host_endian, detail::little_endian>(x);
+}
+
+template <class T>
+T from_little_endian(T x) {
+  return detail::swap<detail::little_endian, detail::host_endian>(x);
+}
+
+segment::header make_header(chunk_ptr chunk) {
+  VAST_ASSERT(chunk->size() >= sizeof(segment::header));
+  auto hdr = reinterpret_cast<const segment::header*>(chunk->data());
+  segment::header result;
+  result.magic = from_little_endian(hdr->magic);
+  result.version = from_little_endian(hdr->version);
+  result.id = hdr->id;
+  result.payload_offset = from_little_endian(hdr->payload_offset);
+  return result;
+}
+
+} // namespace <anonymous>
+
+caf::expected<segment_ptr> segment::make(caf::actor_system& sys, chunk_ptr chunk) {
+  VAST_ASSERT(chunk != nullptr);
+  if (chunk->size() < sizeof(header))
+    return make_error(ec::unspecified, "segment too small", chunk->size());
+  auto hdr = make_header(chunk);
+  if (hdr.magic != magic)
+    return make_error(ec::unspecified, "invalid segment magic", hdr.magic);
+  if (hdr.version > version)
+    return make_error(ec::unspecified, "segment version too big", hdr.version);
+  // Create a segment and copy the header.
+  auto result = caf::make_counted<segment>(sys, chunk);
+  result->header_ = hdr;
+  // Deserialize meta data.
+  caf::charbuf buf{chunk->data(), chunk->size()};
+  detail::coded_deserializer<caf::charbuf&> meta_deserializer{buf};
+  if (auto error = meta_deserializer(result->meta_))
+    return error;
+  return result;
+}
+
+const uuid& segment::id() const {
+  return header_.id;
+}
+
+chunk_ptr segment::chunk() const {
+  return chunk_;
+}
+
+size_t segment::slices() const {
+  return meta_.slices.size();
+}
+
+caf::expected<std::vector<const_table_slice_handle>>
+segment::lookup(const ids& xs) const {
+  std::vector<const_table_slice_handle> result;
+  auto rng = select(xs);
+  auto payload = chunk_->data() + header_.payload_offset;
+  // Walk in lock-step through the slices and the ID sequence.
+  for (auto& slice : meta_.slices) {
+    if (!rng)
+      return result;
+    // Make the ID range catch up if it's behind.
+    if (rng.get() < slice.offset) {
+      rng.skip(slice.offset);
+      if (!rng)
```

Comment:
How can skip invalidate the range? And if so, why has `skip` no return value? Or is `!rng` actually checking for `rng.empty()`? Either way, it's not intuitive to follow what's going on here.

Reply:
I hear you, but your concerns are orthogonal to this PR. Since `skip` is similar to incrementing an iterator, you can get an invalid range after doing so.

> Or is !rng actually checking for rng.empty()?

[Yes](https://github.com/vast-io/vast/blob/master/libvast/vast/detail/range.hpp#L27).